hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/discovery/HazelcastCloudDiscovery.java

```diff
@@ -0,0 +1,135 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.spi.impl.discovery;
+
+import com.eclipsesource.json.Json;
+import com.eclipsesource.json.JsonValue;
+import com.hazelcast.nio.Address;
+import com.hazelcast.util.StringUtil;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Scanner;
+
+import static com.hazelcast.util.ExceptionUtil.rethrow;
+
+/**
+ * Discovery service that discover nodes via hazelcast.cloud
+ * https://coordinator.hazelcast.cloud/cluster/discovery?token=<TOKEN>
+ */
+class HazelcastCloudDiscovery {
+
+    private static final int DEFAULT_TIMEOUT_MILLIS = 30000;
+    private static final int MIN_HTTP_CODE_FOR_ERROR = 400;
+    private static final int MAX_HTTP_CODE_FOR_ERROR = 600;
+    private static final String CLOUD_URL = "https://coordinator.hazelcast.cloud/cluster/discovery?token=";
+    private static final String PRIVATE_ADDRESS_PROPERTY = "private-address";
+    private static final String PUBLIC_ADDRESS_PROPERTY = "public-address";
+
+    private final String endpointUrl;
+
+    HazelcastCloudDiscovery(String cloudToken) {
+        endpointUrl = CLOUD_URL + cloudToken;
+    }
+
+    Map<Address, Address> discoverNodes() {
+        try {
+            return callService();
+        } catch (Exception e) {
+            throw rethrow(e);
+        }
+    }
+
+    private Map<Address, Address> callService() throws IOException, URISyntaxException {
+        URL url = new URL(endpointUrl);
+        HttpURLConnection httpConnection = (HttpURLConnection) url.openConnection();
+        httpConnection.setRequestMethod("GET");
+        httpConnection.setConnectTimeout(DEFAULT_TIMEOUT_MILLIS);
+        httpConnection.setReadTimeout(DEFAULT_TIMEOUT_MILLIS);
+        httpConnection.setRequestProperty("Accept-Charset", "UTF-8");
+        httpConnection.connect();
+        checkError(httpConnection);
+        return parseResponse(httpConnection.getInputStream());
+    }
+
+    private Map<Address, Address> parseResponse(InputStream is) throws IOException, URISyntaxException {
+
+        JsonValue jsonValue = Json.parse(readInputStream(is));
+        List<JsonValue> response = jsonValue.asArray().values();
+
+        Map<Address, Address> privateToPublicAddresses = new HashMap<Address, Address>();
+        for (JsonValue value : response) {
+            String privateAddress = value.asObject().get(PRIVATE_ADDRESS_PROPERTY).asString();
+            String publicAdress = value.asObject().get(PUBLIC_ADDRESS_PROPERTY).asString();
+            publicAdress = StringUtil.isNullOrEmptyAfterTrim(publicAdress) ? privateAddress : publicAdress;
```

Comment:
is this impossible by design of the rest service? do we need this default value?

Reply:
I will remove this line. This is a leftover from old prorotype. We assume that json will always contain public and private address pair.